### PR TITLE
O3-1438: Implement the Schema Editor Component

### PR DESCRIPTION
**Pull request by kumuditha-udayanga**
### Purpose
Implement the schema editor component.

### Approach
- Fetch form schema according to the selected form in the form editor.
- If the form does have a schema form display a template schema on the editor.
- Throws an error is the JSON schema has a syntax error.

#### On create new form `form-builder/edit/new`
Screenshot 2022-07-28 at 9 43 35 PM

#### On a form with schema `form-builder/edit/{form-uuid}`
Screenshot 2022-07-28 at 9 44 08 PM

#### JSON Error
Screenshot 2022-07-28 at 9 45 34 PM

#### On a form with no schema `form-builder/edit/{form-uuid}`
Screenshot 2022-07-28 at 9 45 59 PM

### CODE REVIEW

**Comment by kumuditha-udayanga** [comment]
@samuelmale I made the requested change. Now the error shows inline as follows
Screenshot 2022-08-07 at 9 05 52 PM

File: src/components/form-editor/form-editor.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import { Form, Schema } from "../../api/types";
+
+const FormEditor: React.FC = () => {
+  type State = {
+    uuid: string;
+  };
+  const { t } = useTranslation();
+  const { uuid } = useParams<State>();
+
+  const GetFormMetaData = (formUUID: string) => {
```
Comment (samuelmale) [inline src/components/form-editor/form-editor.tsx:18]: `getFormMetaData`

File: src/components/form-editor/form-editor.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    const { formMetaData, formMetaDataError } = useFormMetadata(formUUID);
```
Comment (samuelmale) [inline src/components/form-editor/form-editor.tsx:19]: Any reason as to why hook `useFormMetadata(formUUID)` is not used at the top level of the `FormEditor` functional component?

File: src/components/form-editor/form-editor.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  const GetRawSchema = () => {
+    const schema: Schema = {
+      name: "",
+      pages: [],
+      processor: "EncounterFormProcessor",
+      uuid: "xxx",
+      referencedForms: [],
+    };
+    return schema;
```
Comment (samuelmale) [inline src/components/form-editor/form-editor.tsx:60]: This abstraction is unnecessary. Just declare this as a constant. I would declare this at the top level of this module (I mean out of the `FormEditor` component scope).

```javascript
import { Form, Schema } from "../../api/types";

const emptyFormSchema = {
      name: "",
      pages: [],
      processor: "EncounterFormProcessor",
      uuid: "xxx",
      referencedForms: [],
    };

const FormEditor: React.FC = () => { 
```

File: src/components/form-editor/schema-editor/schema-editor.component.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import React, { useCallback, useEffect, useState } from "react";
+import { Button } from "carbon-components-react";
+import styles from "./schema-editor.scss";
+import AceEditor from "react-ace";
+import { useTranslation } from "react-i18next";
+
+interface SchemaEditorProps {
+  schema: any;
```
Comment (samuelmale) [inline src/components/form-editor/schema-editor/schema-editor.component.tsx:8]: Typing?

File: src/components/form-editor/schema-editor/schema-editor.component.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  const { t } = useTranslation();
+  const [formSchema, setFormSchema] = useState<string>("");
+  const [errorMessage, setErrorMessage] = useState("");
+
+  const updateFormJson = useCallback(
+    (value) => {
+      setFormSchema(value);
+    },
+    [setFormSchema]
+  );
```
Comment (samuelmale) [inline src/components/form-editor/schema-editor/schema-editor.component.tsx:21]: What stops you from calling `setFormSchema(value)` directly?

File: src/index.ts [hunk 6]
```diff
@@ -14,6 +14,8 @@ const backendDependencies = {
 
 function setupOpenMRS() {
   const moduleName = "@openmrs/esm-form-builder";
+  const route = `form-builder`;
+  const spaBasePath = `${window.spaBase}/${route}`;
```
Comment (samuelmale) [inline src/index.ts:18]: If you aren't re-using the `route` variable anywhere else, just:

```js
const spaBasePath = `${window.spaBase}/form-builder`;
```

**Review by samuelmale — CHANGES_REQUESTED** [review]
Looks good, just a few minor requests.

File: src/components/form-editor/form-editor.tsx (hunk 2 above)
Comment (kumuditha-udayanga) [reply src/components/form-editor/form-editor.tsx:19]: The reason is in a scenario of `Create new form` this wont be needed.

Comment (samuelmale) [reply src/components/form-editor/form-editor.tsx:19]: I would advise invoking that from the top. I don't see anything special that prevents that..

Comment (kumuditha-udayanga) [reply src/components/form-editor/form-editor.tsx:19]: Sure

File: src/index.ts (hunk 6 above)
Comment (kumuditha-udayanga) [reply src/index.ts:18]: I have used `route` in line number 29

**Comment by kumuditha-udayanga** [comment]
@samuelmale, I made the requested changes

**Pull request merged**